pkg/gen/python-templates/CustomResource.py

```diff
@@ -15,37 +19,74 @@ class CustomResource(pulumi.CustomResource):
-        if not isinstance(__name__, str):
+        if not isinstance(resource_name, str):
             raise TypeError('Expected resource name to be a string')
-        if __opts__ and not isinstance(__opts__, pulumi.ResourceOptions):
+        if opts and not isinstance(opts, pulumi.ResourceOptions):
             raise TypeError('Expected resource options to be a ResourceOptions instance')
 
         __props__ = dict()
 
-        if not api_version:
-            raise TypeError('Missing required property apiVersion')
         __props__['apiVersion'] = api_version
-        if not kind:
-            raise TypeError('Missing required property kind')
         __props__['kind'] = kind
-        if spec is None:
-            raise TypeError('Missing required property spec')
         __props__['spec'] = spec
         __props__['metadata'] = metadata
 
-        if __opts__ is None:
-            __opts__ = pulumi.ResourceOptions()
-        if __opts__.version is None:
-            __opts__.version = version.get_version()
+        if opts is None:
+            opts = pulumi.ResourceOptions()
+        if opts.version is None:
+            opts.version = version.get_version()
 
         super(CustomResource, self).__init__(
-            "kubernetes:%s:%s" % (api_version, kind),
-            __name__,
+            f"kubernetes:{api_version}:{kind}",
+            resource_name,
             __props__,
-            __opts__)
+            opts)
+
+    @staticmethod
+    def get(resource_name: str,
+            api_version: pulumi.Input[str],
+            kind: pulumi.Input[str],
+            id: pulumi.Input[str],
+            opts: Optional[ResourceOptions] = None):
+        """
+        :param str resource_name: _Unique_ name used to register this resource with Pulumi.
+        :param pulumi.Input[str] api_version: The API version of the apiExtensions.CustomResource we
+               wish to select, as specified by the CustomResourceDefinition that defines it on the
+               API server.
+        :param pulumi.Input[str] kind: The kind of the apiextensions.CustomResource we wish to select,
+               as specified by the CustomResourceDefinition that defines it on the API server.
+        :param pulumi.Input[str] id: An ID for the Kubernetes resource to retrieve.
+               Takes the form <namespace>/<name> or <name>.
+        :param Optional[ResourceOptions] opts: A bag of options that control this resource's behavior.
+        """
+        _api_version = pulumi.Output.from_input(api_version)
+        _kind = pulumi.Output.from_input(kind)
+        _id = pulumi.Output.from_input(id)
+
+        def _unwrap(all_config):
+            api_version, kind, id = all_config
+            opts = ResourceOptions(id=id) if opts is None else opts.merge(ResourceOptions(id=id))
+            return CustomResource(resource_name=resource_name, api_version=api_version, kind=kind, opts=opts)
```

Comment:
this seems shady.   your constructor above already accepts:

```
            api_version: pulumi.Input[str],
            kind: pulumi.Input[str],
```

So no need to coerce those to outputs.  The main issue seems to be 'id'.  However, walking down to the core pulumi lib, ti looks like we should be able to support any `Input[str]` as an id.

Reply:
I realized that `api_version` and `kind` really didn't need to be Inputs at all. It looks like `ResourceOptions` doesn't allow an Input for `id`:
https://github.com/pulumi/pulumi/blob/20058f982d3259f1d55bc1d081bc3557d24ef920/sdk/python/lib/pulumi/resource.py#L274